File: packages/python-sdk/src/blindfold/regex/scanner.py
"""PIIScanner: orchestrates detectors, deduplicates results, produces output"""

import base64
import hashlib
import os
from typing import Dict, List, Optional, Tuple

from .entities import EntityType, PIIMatch
from .registry import DetectorRegistry
from .synthesizers import synthesize_value

# Force import of all detectors so they auto-register
from . import detectors as _detectors  # noqa: F401


class TokenizeResult:
    """Result of tokenize operation."""

    __slots__ = ("text", "mapping", "matches")

    def __init__(self, text: str, mapping: Dict[str, str], matches: List[PIIMatch]) -> None:
        self.text = text
        self.mapping = mapping
        self.matches = matches


class MaskResult:
    """Result of mask operation."""

    __slots__ = ("text", "matches")

    def __init__(self, text: str, matches: List[PIIMatch]) -> None:
        self.text = text
        self.matches = matches


class HashResult:
    """Result of hash operation."""

    __slots__ = ("text", "matches")

    def __init__(self, text: str, matches: List[PIIMatch]) -> None:
        self.text = text
        self.matches = matches


class EncryptResult:
    """Result of encrypt operation."""

    __slots__ = ("text", "matches")

    def __init__(self, text: str, matches: List[PIIMatch]) -> None:
        self.text = text
        self.matches = matches


class SynthesizeResult:
    """Result of synthesize operation."""

    __slots__ = ("text", "matches")

    def __init__(self, text: str, matches: List[PIIMatch]) -> None:
        self.text = text
        self.matches = matches


class PIIScanner:
    """Local regex-based PII scanner.

    Usage::

        scanner = PIIScanner(locales=["us", "eu"])
        matches = scanner.detect("Email john@acme.com, SSN 123-45-6789")
        redacted, matches = scanner.redact("Email john@acme.com")

    Args:
        locales: List of locale codes to enable (default ``["us"]``).
    """

    def __init__(
        self,
        locales: Optional[List[str]] = None,
    ) -> None:
        self._registry = DetectorRegistry(locales=locales)

    def detect(
        self,
        text: str,
        entities: Optional[List[str]] = None,
    ) -> List[PIIMatch]:
        """Detect PII entities in *text* and return deduplicated matches.

        Args:
            text: Text to scan.
            entities: Optional list of entity type names to restrict detection.
        """
        detectors = self._registry.detectors
        if entities:
            entity_set = set(entities)
            detectors = [d for d in detectors if d.entity_type.value in entity_set]
        has_digit = any(c.isdigit() for c in text)
        text_lower = text.lower()
        all_matches: List[PIIMatch] = []
        for detector in detectors:
            if detector.needs_digit and not has_digit:
                continue
            detector._text_lower = text_lower
            all_matches.extend(detector.iter_matches(text))
            detector._text_lower = None
        return self._deduplicate(all_matches)

    def redact(
        self,
        text: str,
        entities: Optional[List[str]] = None,
    ) -> Tuple[str, List[PIIMatch]]:
        """Detect and remove PII from text.

        Returns a tuple of (redacted_text, detected_matches).
        """
        matches = self.detect(text, entities=entities)
        if not matches:
            return text, []

        # Sort by start position descending so replacements don't shift indices
        sorted_matches = sorted(matches, key=lambda m: m.start, reverse=True)
        result = text
        for m in sorted_matches:
            # Also consume a preceding space to keep output clean
            start = m.start - 1 if m.start > 0 and result[m.start - 1] == " " else m.start
            result = result[:start] + result[m.end:]

        return result, matches

    def tokenize(self, text: str, entities: Optional[List[str]] = None) -> TokenizeResult:
        """Detect PII and replace with numbered tokens like ``<Email Address_1>``.

        Returns a :class:`TokenizeResult` with tokenized text, mapping, and matches.
        """
        matches = self.detect(text, entities=entities)
        if not matches:
            return TokenizeResult(text=text, mapping={}, matches=[])

        # Assign per-type counters in reading order (ascending start position)
        sorted_asc = sorted(matches, key=lambda m: m.start)
        counters: Dict[str, int] = {}
        token_map: Dict[int, str] = {}  # match id -> token
        mapping: Dict[str, str] = {}

        for m in sorted_asc:
            count = counters.get(m.entity_type, 0) + 1
            counters[m.entity_type] = count
            token = f"<{m.entity_type}_{count}>"
            token_map[id(m)] = token
            mapping[token] = m.text

        # Replace in descending position order
        sorted_desc = sorted(matches, key=lambda m: m.start, reverse=True)
        result = text
        for m in sorted_desc:
            token = token_map[id(m)]
            result = result[:m.start] + token + result[m.end:]

        return TokenizeResult(text=result, mapping=mapping, matches=matches)

    def mask(
        self,
        text: str,
        chars_to_show: int = 3,
        from_end: bool = False,
        masking_char: str = "*",
        entities: Optional[List[str]] = None,
    ) -> MaskResult:
        """Detect PII and partially mask each match."""
        matches = self.detect(text, entities=entities)
        if not matches:
            return MaskResult(text=text, matches=[])

        sorted_matches = sorted(matches, key=lambda m: m.start, reverse=True)
        result = text
        for m in sorted_matches:
            original = m.text
            show = min(chars_to_show, len(original))
            mask_len = len(original) - show
            if from_end:
                masked = masking_char * mask_len + original[mask_len:]
            else:
                masked = original[:show] + masking_char * mask_len
            result = result[:m.start] + masked + result[m.end:]

        return MaskResult(text=result, matches=matches)

    def hash(
        self,
        text: str,
        hash_type: str = "sha256",
        hash_prefix: str = "HASH_",
        hash_length: int = 16,
        entities: Optional[List[str]] = None,
    ) -> HashResult:
        """Detect PII and replace each match with a deterministic hash."""
        matches = self.detect(text, entities=entities)
        if not matches:
            return HashResult(text=text, matches=[])

        sorted_matches = sorted(matches, key=lambda m: m.start, reverse=True)
        result = text
        for m in sorted_matches:
            h = hashlib.new(hash_type)
            h.update(m.text.encode("utf-8"))
            digest = h.hexdigest()[:hash_length]
            replacement = hash_prefix + digest
            result = result[:m.start] + replacement + result[m.end:]

        return HashResult(text=result, matches=matches)

    def encrypt(self, text: str, encryption_key: str, entities: Optional[List[str]] = None) -> EncryptResult:
        """Detect PII and encrypt each match with AES-256-CBC.

        Args:
            text: Text to process.
            encryption_key: Required, minimum 16 characters.

        Raises:
            ValueError: If *encryption_key* is too short.
            ImportError: If the ``cryptography`` package is not installed.
        """
        if not encryption_key or len(encryption_key) < 16:
            raise ValueError(
                "encryption_key is required and must be at least 16 characters for local encryption"
            )

        try:
            from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
            from cryptography.hazmat.primitives import padding as sym_padding
        except ImportError:
            raise ImportError(
                "The 'cryptography' package is required for local encryption. "
                "Install it with: pip install cryptography"
            )

        matches = self.detect(text, entities=entities)
        if not matches:
            return EncryptResult(text=text, matches=[])

        # Derive a 32-byte key via SHA-256
        derived_key = hashlib.sha256(encryption_key.encode("utf-8")).digest()

        sorted_matches = sorted(matches, key=lambda m: m.start, reverse=True)
        result = text
        for m in sorted_matches:
            iv = os.urandom(16)
            cipher = Cipher(algorithms.AES(derived_key), modes.CBC(iv))
            encryptor = cipher.encryptor()
            padder = sym_padding.PKCS7(128).padder()
            padded = padder.update(m.text.encode("utf-8")) + padder.finalize()
            ct = encryptor.update(padded) + encryptor.finalize()
            combined = iv + ct
            replacement = base64.b64encode(combined).decode("ascii")
            result = result[:m.start] + replacement + result[m.end:]

        return EncryptResult(text=result, matches=matches)

    def synthesize(self, text: str, language: Optional[str] = None, entities: Optional[List[str]] = None) -> SynthesizeResult:
        """Detect PII and replace each match with format-preserving synthetic data.

        Args:
            text: Text to process.
            language: Accepted for API compatibility but ignored locally.
        """
        matches = self.detect(text, entities=entities)
        if not matches:
            return SynthesizeResult(text=text, matches=[])

        sorted_matches = sorted(matches, key=lambda m: m.start, reverse=True)
        result = text
        for m in sorted_matches:
            replacement = synthesize_value(m.entity_type, m.text)
            result = result[:m.start] + replacement + result[m.end:]

        return SynthesizeResult(text=result, matches=matches)

    @staticmethod
    def _deduplicate(matches: List[PIIMatch]) -> List[PIIMatch]:
        """Remove overlapping matches, preferring higher score then longer span."""
        if not matches:
            return []
        # Sort by start ascending, then by score descending, then span length descending
        matches.sort(key=lambda m: (m.start, -m.score, -(m.end - m.start)))
        result: List[PIIMatch] = []
        last_end = -1
        for m in matches:
            if m.start >= last_end:
                result.append(m)
                last_end = m.end
        return result

File: packages/js-sdk/src/regex/scanner.ts
// PIIScanner: orchestrates detectors, deduplicates results, produces output

import * as crypto from 'crypto'
import { EntityType, PIIMatch } from './entities'
import { DetectorRegistry } from './registry'
import { synthesizeValue } from './synthesizers'

// Force import of all detectors so they auto-register
import './detectors/index'

export interface PIIScannerOptions {
  /** List of locale codes to enable (default: ["us"]). */
  locales?: string[]
}

export interface TokenizeResult {
  text: string
  mapping: Record<string, string>
  matches: PIIMatch[]
}

export interface MaskResult {
  text: string
  matches: PIIMatch[]
}

export interface HashResult {
  text: string
  matches: PIIMatch[]
}

export interface EncryptResult {
  text: string
  matches: PIIMatch[]
}

export interface SynthesizeResult {
  text: string
  matches: PIIMatch[]
}

/**
 * Local regex-based PII scanner.
 *
 * Usage:
 *   const scanner = new PIIScanner({ locales: ['us', 'eu'] })
 *   const matches = scanner.detect('Email john@acme.com, SSN 123-45-6789')
 *   const [redacted, matches] = scanner.redact('Email john@acme.com')
 */
export class PIIScanner {
  private registry: DetectorRegistry

  constructor(options?: PIIScannerOptions) {
    this.registry = new DetectorRegistry(options?.locales)
  }

  /** Detect PII entities in text and return deduplicated matches. */
  detect(text: string, entities?: string[]): PIIMatch[] {
    let detectors = this.registry.detectors
    if (entities) {
      const entitySet = new Set(entities)
      detectors = detectors.filter((d) => entitySet.has(d.entityType))
    }
    const hasDigit = /\d/.test(text)
    const textLower = text.toLowerCase()
    const allMatches: PIIMatch[] = []
    for (const detector of detectors) {
      if (detector.needsDigit && !hasDigit) {
        continue
      }
      ;(detector as any)._textLower = textLower
      allMatches.push(...detector.iterMatches(text))
      ;(detector as any)._textLower = null
    }
    return this.deduplicate(allMatches)
  }

  /**
   * Detect and remove PII from text.
   * Returns a tuple of [redactedText, detectedMatches].
   */
  redact(text: string, entities?: string[]): [string, PIIMatch[]] {
    const matches = this.detect(text, entities)
    if (matches.length === 0) {
      return [text, []]
    }

    // Sort by start position descending so replacements do not shift indices
    const sortedMatches = [...matches].sort((a, b) => b.start - a.start)
    let result = text
    for (const m of sortedMatches) {
      // Also consume a preceding space to keep output clean
      const start = m.start > 0 && result[m.start - 1] === ' ' ? m.start - 1 : m.start
      result = result.slice(0, start) + result.slice(m.end)
    }

    return [result, matches]
  }

  /**
   * Detect PII and replace with numbered tokens like `<Email Address_1>`.
   * Returns tokenized text, a mapping from tokens to original values, and matches.
   */
  tokenize(text: string, entities?: string[]): TokenizeResult {
    const matches = this.detect(text, entities)
    if (matches.length === 0) {
      return { text, mapping: {}, matches: [] }
    }

    // Assign per-type counters in reading order (ascending start position)
    const sortedAsc = [...matches].sort((a, b) => a.start - b.start)
    const counters: Record<string, number> = {}
    const tokenMap: Map<PIIMatch, string> = new Map()
    const mapping: Record<string, string> = {}

    for (const m of sortedAsc) {
      const count = (counters[m.entityType] ?? 0) + 1
      counters[m.entityType] = count
      const token = `<${m.entityType}_${count}>`
      tokenMap.set(m, token)
      mapping[token] = m.text
    }

    // Replace in descending position order
    const sortedDesc = [...matches].sort((a, b) => b.start - a.start)
    let result = text
    for (const m of sortedDesc) {
      const token = tokenMap.get(m)!
      result = result.slice(0, m.start) + token + result.slice(m.end)
    }

    return { text: result, mapping, matches }
  }

  /**
   * Detect PII and partially mask each match.
   */
  mask(
    text: string,
    charsToShow: number = 3,
    fromEnd: boolean = false,
    maskingChar: string = '*',
    entities?: string[]
  ): MaskResult {
    const matches = this.detect(text, entities)
    if (matches.length === 0) {
      return { text, matches: [] }
    }

    const sortedMatches = [...matches].sort((a, b) => b.start - a.start)
    let result = text
    for (const m of sortedMatches) {
      const original = m.text
      const show = Math.min(charsToShow, original.length)
      const maskLen = original.length - show
      let masked: string
      if (fromEnd) {
        masked = maskingChar.repeat(maskLen) + original.slice(maskLen)
      } else {
        masked = original.slice(0, show) + maskingChar.repeat(maskLen)
      }
      result = result.slice(0, m.start) + masked + result.slice(m.end)
    }

    return { text: result, matches }
  }

  /**
   * Detect PII and replace each match with a deterministic hash.
   */
  hash(
    text: string,
    hashType: string = 'sha256',
    hashPrefix: string = 'HASH_',
    hashLength: number = 16,
    entities?: string[]
  ): HashResult {
    const matches = this.detect(text, entities)
    if (matches.length === 0) {
      return { text, matches: [] }
    }

    const sortedMatches = [...matches].sort((a, b) => b.start - a.start)
    let result = text
    for (const m of sortedMatches) {
      const digest = crypto.createHash(hashType).update(m.text).digest('hex')
      const truncated = digest.slice(0, hashLength)
      const replacement = hashPrefix + truncated
      result = result.slice(0, m.start) + replacement + result.slice(m.end)
    }

    return { text: result, matches }
  }

  /**
   * Detect PII and encrypt each match with AES-256-CBC.
   * @param encryptionKey - Required, minimum 16 characters.
   */
  encrypt(text: string, encryptionKey: string, entities?: string[]): EncryptResult {
    if (!encryptionKey || encryptionKey.length < 16) {
      throw new Error('encryptionKey is required and must be at least 16 characters for local encryption')
    }

    const matches = this.detect(text, entities)
    if (matches.length === 0) {
      return { text, matches: [] }
    }

    // Derive a 32-byte key from the provided key using SHA-256
    const derivedKey = crypto.createHash('sha256').update(encryptionKey).digest()

    const sortedMatches = [...matches].sort((a, b) => b.start - a.start)
    let result = text
    for (const m of sortedMatches) {
      const iv = crypto.randomBytes(16)
      const cipher = crypto.createCipheriv('aes-256-cbc', derivedKey, iv)
      const encrypted = Buffer.concat([cipher.update(m.text, 'utf8'), cipher.final()])
      const combined = Buffer.concat([iv, encrypted])
      const replacement = combined.toString('base64')
      result = result.slice(0, m.start) + replacement + result.slice(m.end)
    }

    return { text: result, matches }
  }

  /**
   * Detect PII and replace each match with format-preserving synthetic data.
   * @param _language - Accepted for API compatibility but ignored locally.
   */
  synthesize(text: string, _language?: string, entities?: string[]): SynthesizeResult {
    const matches = this.detect(text, entities)
    if (matches.length === 0) {
      return { text, matches: [] }
    }

    const sortedMatches = [...matches].sort((a, b) => b.start - a.start)
    let result = text
    for (const m of sortedMatches) {
      const replacement = synthesizeValue(m.entityType, m.text)
      result = result.slice(0, m.start) + replacement + result.slice(m.end)
    }

    return { text: result, matches }
  }

  /** Remove overlapping matches, preferring higher score then longer span. */
  private deduplicate(matches: PIIMatch[]): PIIMatch[] {
    if (matches.length === 0) {
      return []
    }
    // Sort by start ascending, then by score descending, then span length descending
    matches.sort((a, b) => {
      if (a.start !== b.start) return a.start - b.start
      if (a.score !== b.score) return b.score - a.score
      return b.end - b.start - (a.end - a.start)
    })
    const result: PIIMatch[] = []
    let lastEnd = -1
    for (const m of matches) {
      if (m.start >= lastEnd) {
        result.push(m)
        lastEnd = m.end
      }
    }
    return result
  }
}
